xdsl/transforms/memref_stream_legalize.py

```diff
@@ -0,0 +1,163 @@
+from dataclasses import dataclass
+from typing import TypeAlias, cast
+
+from xdsl.context import MLContext
+from xdsl.dialects import memref_stream
+from xdsl.dialects.builtin import (
+    ArrayAttr,
+    Float16Type,
+    Float32Type,
+    Float64Type,
+    IntegerAttr,
+    ModuleOp,
+    VectorType,
+    f16,
+    f32,
+)
+from xdsl.dialects.linalg import IteratorType
+from xdsl.ir import Attribute, Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.exceptions import DiagnosticException
+from xdsl.utils.hints import isa
+
+StreamingVectorType: TypeAlias = VectorType[Float64Type | Float32Type | Float16Type]
+StreamingLegalType: TypeAlias = StreamingVectorType | Float64Type
+
+
+def _is_legal_vector(attr: Attribute) -> bool:
+    if isa(attr, StreamingVectorType):
+        match attr.element_type, attr.element_count():
+            case Float64Type(), 1:
+                return True
+            case Float32Type(), 2:
+                return True
+            case Float16Type(), 4:
+                return True
+            case _:
+                return False
+    return False
+
+
+def _is_legal_attr(attr: Attribute) -> bool:
+    return _is_legal_vector(attr) or isa(attr, StreamingLegalType)
```

Comment:
It seems like these helpers should be on the memref_stream.generic type itself, or maybe in the dialect definition file?

Reply:
Actually no they're snitch constraints not memref_stream constraints